specs/simple-serialize.md

```diff
@@ -1,424 +1,141 @@
-
-To serialize:
-
-1. Get the list of the container's fields.
-
-2. For each name in the list, obtain the corresponding value from the container and serialize it. Place this serialized value into a buffer. The serialized values should be tightly packed.
-
-3. Get the number of raw bytes in the serialized buffer. Encode that number as a `4-byte` **little endian** `uint32`.
-
-4. Prepend the length to the serialized buffer.
-
-**Example in Python**
+#### Containers
 
 ```python
-def get_field_names(typ):
-    return typ.fields.keys()
-
-def get_value_for_field_name(value, field_name):
-    return getattr(value, field_name)
-
-def get_type_for_field_name(typ, field_name):
-    return typ.fields[field_name]
-
-serialized_buffer = b''
-
-typ = type(value)
-for field_name in get_field_names(typ):
-    field_value = get_value_for_field_name(value, field_name)
-    field_type = get_type_for_field_name(typ, field_name)
-    serialized_buffer += serialize(field_value, field_type)
-
-assert(len(serialized_buffer) < 2**32)
-
-serialized_len = (len(serialized_buffer).to_bytes(LENGTH_BYTES, 'little'))
-
-return serialized_len + serialized_buffer
+serialized_elements = [serialize(element) for element in object]
+serialized_bytes = reduce(lambda x, y: x + y, serialized_elements)
+assert len(serialized_bytes) < MAX_LENGTH
+serialized_length = len(serialized_bytes).to_bytes(LENGTH_BYTES, 'little')
+return serialized_length + serialized_bytes
 ```
 
-### Deserialize/Decode
-
-The decoding requires knowledge of the type of the item to be decoded. When
-performing decoding on an entire serialized string, it also requires knowledge
-of the order in which the objects have been serialized.
-
-Note: Each return will provide:
-- `deserialized_object`
-- `new_index`
-
-At each step, the following checks should be made:
-
-| Check to perform         | Check                                                      |
-|:-------------------------|:-----------------------------------------------------------|
-| Ensure sufficient length | ``len(rawbytes) >= current_index + deserialize_length``    |
-
-At the final step, the following checks should be made:
-
-| Check to perform         | Check                                |
-|:-------------------------|:-------------------------------------|
-| Ensure no extra length   | `new_index == len(rawbytes)`         |
-
-#### uintN
-
-Convert directly from bytes into integer utilising the number of bytes the same
-size as the integer length. (e.g. ``uint16 == 2 bytes``)
-
-All integers are interpreted as **little endian**.
+#### Tuples
 
 ```python
-byte_length = int_size / 8
-new_index = current_index + byte_length
-assert(len(rawbytes) >= new_index)
-return int.from_bytes(rawbytes[current_index:current_index+byte_length], 'little'), new_index
+serialized_elements = [serialize(element) for element in object]
+serialized_bytes = reduce(lambda x, y: x + y, serialized_elements)
+return serialized_bytes
```

Comment:
Why don't tuples have a length prefix? I thought that maybe the reason is that it's not necessary because one knows the number of elements in advance, but the same is true for containers which do have the prefix.

Reply:
> the reason is that it's not necessary because one knows the number of elements in advance

Right! I think the idea is that developers can free up a buffer of the right size when reading data on the wire.

> the same is true for containers which do have the prefix

For containers `len(serialized_bytes)` is not always known ahead of time.

It's a bit awkward that length is handled in three different ways for contains, lists and tuples. Maybe we can just use `len(serialized_bytes)` everywhere.

Reply:
@jannikluhn: I've gone ahead and used the exact same serialisation logic for containers, lists and tuples. It's slightly less efficient but goes well with the "Simple" theme of SSZ.

cc @djrtwo, @vbuterin

Reply:
Thanks @jannikluhn, @djrtwo, @vbuterin for the reviews. All comments have been addressed :)